AI Chat: show web sources that were used via search to generate the response

petemill opened the pull request:
<!-- Add brave-browser issue below that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/39630

<!-- CI-related labels that can be applied to this PR:
* CI/run-audit-deps (1) - check for known npm/cargo vulnerabilities (audit_deps)
* CI/run-network-audit (1) - run network-audit
* CI/run-upstream-tests - run Chromium unit and browser tests on Linux and Windows (otherwise only on Linux)
* CI/run-linux-arm64, CI/run-macos-x64, CI/run-windows-arm64, CI/run-windows-x86 - run builds that would otherwise be skipped
* CI/skip - do not run CI builds (except noplatform)
* CI/skip-linux-x64, CI/skip-android, CI/skip-macos-arm64, CI/skip-ios, CI/skip-windows-x64 - skip CI builds for specific platforms
* CI/skip-upstream-tests - do not run Chromium unit, or browser tests (otherwise only on Linux)
* CI/skip-all-linters - do not run presubmit and lint checks
* CI/storybook-url (1) - deploy storybook and provide a unique URL for each build

(1) applied automatically when some files are changed (see: https://github.com/brave/brave-core/blob/master/.github/labeler.yml)
-->

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

diracdeltas commented on components/ai_chat/resources/page/components/assistant_response/web_sources_event.tsx:
```diff
@@ -0,0 +1,62 @@
+          const host = new URL(source.url.url).hostname
+          return (
+            <li>
+              <a href={source.url.url} title={source.title} onClick={(e) => handleOpenSource(e, source)}>
```
can you add some validation to make sure `source.url.url` has a valid protocol (`http` or `https`)?

petemill replied:
This _is_ handled on the C++ end at https://github.com/brave/brave-core/blob/d2863592ec75dd0ae15c15e8b599c89070ae6b98/browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc#L263 but I'll also not make it a link here

thypon commented on browser/ui/webui/ai_chat/ai_chat_ui_page_handler.h:
```diff
@@ -37,6 +38,10 @@ namespace favicon {
 }  // namespace favicon
 
+namespace network {
+class SimpleURLLoader;
```
why we need an extra favicon url loader?

petemill replied:
We can't load the favicon from the internal favicon database because the user might not have visited the site before, if that's what you mean?

thypon commented on components/ai_chat/core/browser/engine/conversation_api_client_unittest.cc:
```diff
@@ -217,6 +217,27 @@ TEST_F(ConversationAPIUnitTest, PerformRequest_PremiumHeaders) {
+            query.GetDict().Set("url", "https://startrek.com");
+            query.GetDict().Set("favicon", "https://startrek.com/favicon");
+            sources.GetList().Append(std::move(query));
+          }
```
this should probably fail for `http://` favicons and `javascript:...` URLs

petemill replied:
The parsing will succeed but we won't link to the site or download the favicon

petemill commented:
Converted to a draft as we decided to refactor quite a bit of the trusted / untrusted boundary to allow for images to be loaded directly

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://build-artifacts.brave.com/brave-browser-build-pr/aichat-sources/storybook.html)

diracdeltas commented on components/ai_chat/core/browser/engine/conversation_api_client.cc:
```diff
@@ -114,6 +114,35 @@ mojom::ConversationEntryEventPtr ParseResponseEvent(
+      }
+      GURL item_url(*url);
+      GURL item_favicon_url(*favicon_url);
+      if (!item_url.is_valid() || !item_favicon_url.is_valid()) {
```
i thought we were only loading favicons from imgs.search.brave.com? can we add that as a host check before loading the image?

petemill replied:
Yes I suppose we have to move the restriction to here now if we still want it, since the fetch is happening more generically and can't be enforced at the point of download.

diracdeltas replied:
let's keep the restriction for now

petemill replied:
done

bridiver commented on components/ai_chat/core/browser/engine/conversation_api_client.cc:
```diff
@@ -114,6 +117,47 @@ mojom::ConversationEntryEventPtr ParseResponseEvent(
     }
     return mojom::ConversationEntryEvent::NewSearchQueriesEvent(
         std::move(event));
+  } else if (*type == "webSources") {
```
for follow-up these should be constants in mojo

bridiver commented on components/ai_chat/core/browser/engine/conversation_api_client.cc:
```diff
@@ -114,6 +117,47 @@ mojom::ConversationEntryEventPtr ParseResponseEvent(
+    if (!sources) {
+      return nullptr;
+    }
+    auto event = mojom::WebSourcesEvent::New();
```
tests for parsing these?

bridiver replied:
nvm, I see it now

bridiver commented on components/ai_chat/core/browser/engine/conversation_api_client_unittest.cc:
```diff
@@ -308,6 +352,21 @@ TEST_F(ConversationAPIUnitTest, PerformRequest_PremiumHeaders) {
         EXPECT_EQ(queries[0], "Star Wars");
         EXPECT_EQ(queries[1], "Star Trek");
       });
+  EXPECT_CALL(mock_callbacks, OnDataReceived(_))
```
what does it look like when something in here fails? While this may be convenient, I'm a bit concerned that failures will be very hard to understand and trace back

thypon commented on components/ai_chat/resources/untrusted_conversation_frame/components/assistant_response/web_sources_event.tsx:
```diff
@@ -0,0 +1,61 @@
+  return (
+    <li>
+      <a href={source.url.url} title={source.title} onClick={(e) => handleOpenSource(e, source)}>
+        <img src={createSanitizedImageUrl(source.faviconUrl.url)} />
```
Am I understanding right that we are loading the favicon as an `chrome-untrusted://image?src=...` here?
Do we have a proxied database of favicons we can get from `search`?

petemill replied:
I'm not sure I understand what you're asking @thypon. We are using the Search API which is returning sources used in the search. Each source has a favicon url. We are passing that favicon url to the image sanitizer here.

thypon replied:
rephrasing: are these favicon the ones proxied by Brave Search itself, or are the original favicon url of the website?
In the second case we are introducing a leak through favicon download.

petemill replied:
The API is only returning proxied or pre-cached urls from the search API directly. The browser is only allowing imgs.search.brave.com items.

thypon commented on browser/ui/webui/ai_chat/ai_chat_untrusted_conversation_ui.cc:
```diff
@@ -67,6 +71,16 @@ class UIHandler : public ai_chat::mojom::UntrustedUIHandler {
+    if (!url.is_valid() || !url.SchemeIs(url::kHttpsScheme)) {
+      return;
+    }
+    OpenURL(url);
```
What's the `Referer` in this case?

thypon commented on browser/ui/webui/untrusted_sanitized_image_source.cc:
```diff
@@ -0,0 +1,35 @@
+    const GURL& url,
+    const content::WebContents::Getter& wc_getter,
+    content::URLDataSource::GotDataCallback callback) {
+  if (!url.is_valid() || !url.SchemeIs(content::kChromeUIUntrustedScheme)) {
```
I'm not super sure we want to download a favicon, before we actually visited the website, here.

diracdeltas commented on .storybook/web-common-mock/create_sanitized_image_url.ts:
```diff
@@ -0,0 +1,8 @@
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+export default function createSanitizedImageUrl(imageUrl: string) {
+  return imageUrl
```
any reason to not encodeURIcomponent here?

diracdeltas replied:
nvm, i see this is not doing any sanitization

bridiver commented on browser/ui/BUILD.gn:
```diff
@@ -101,6 +101,8 @@ source_set("ui") {
     "webui/brave_webui_source.h",
     "webui/skus_internals_ui.cc",
     "webui/skus_internals_ui.h",
+    "webui/untrusted_sanitized_image_source.cc",
```
can we put these in a separate target for `///brave/browser/ui:untrusted_sanitized_image_source`?

bridiver replied:
actually maybe that doesn't help anything with the current state of things so it's ok for now

petemill commented:
Please take another look, especially @bridiver who asked for a browser test:
- Added a frontend test (had to change some jest config to get this working. I am [experimenting with upgrading jest entirely in another branch](https://github.com/brave/brave-core/compare/master..aichat-sources-jest), but will save that for a follow-up and as time allows). I think this negates the need for a browser test since we are testing the parsing and the rendering now.
- Added persistence to the SQL database for these events, including unit tests. This is done via converting to protobuf and then serializing to string and encrypting before storing the array of web sources in a single BLOB column.

diracdeltas commented:
> * Added persistence to the SQL database for these events, including unit tests. This is done via converting to protobuf and then serializing to string and encrypting before storing the array of web sources in a single BLOB column.

is this just to support showing favicons/urls in persisted conversations? it needs to follow the privacy constraints on those conversations if so (deleted with history, can opt out, etc.). cc @ShivanKaul

petemill commented:
> > * Added persistence to the SQL database for these events, including unit tests. This is done via converting to protobuf and then serializing to string and encrypting before storing the array of web sources in a single BLOB column.
> 
> is this just to support showing favicons/urls in persisted conversations? it needs to follow the privacy constraints on those conversations if so (deleted with history, can opt out, etc.). cc @ShivanKaul

This is for web sources that the LLM used to generate any response. So I'm not sure the connection with deleting browser history. For that we only delete the browser history and web content association with a conversation, but not the conversation content or LLM response, the latter of which I believe this falls under.

Of course if a conversation is deleted then all conversation entry content will be deleted, including the web sources.

ShivanKaul commented:
> is this just to support showing favicons/urls in persisted conversations? it needs to follow the privacy constraints on those conversations if so (deleted with history, can opt out, etc.). cc @ShivanKaul

We don't delete anything from the content of the conversations, so it makes sense to me to not do that here either, given that the favicons and URLs are just considered part of the chat's contents.

diracdeltas commented:
@petemill i was specifically referring to the sql persistence of web source events (https://github.com/brave/brave-core/pull/24560/commits/e83b606f9ef72067257af8f5458293b4f00d9708#diff-2b5ac683e368da59575e8bd602ea4c4642efe278d707444d393c32ca072a1c7cR744) since those include url and favicon url. i guess it makes sense not to delete those with browsing history since it's not related, but want to make sure if a user has not opted into conversation history or deletes conversation history, this stuff is also deleted. sounds like it is.

petemill commented:
> @petemill i was specifically referring to the sql persistence of web source events ([e83b606#diff-2b5ac683e368da59575e8bd602ea4c4642efe278d707444d393c32ca072a1c7cR744](https://github.com/brave/brave-core/commit/e83b606f9ef72067257af8f5458293b4f00d9708#diff-2b5ac683e368da59575e8bd602ea4c4642efe278d707444d393c32ca072a1c7cR744)) since those include url and favicon url. i guess it makes sense not to delete those with browsing history since it's not related, but want to make sure if a user has not opted into conversation history or deletes conversation history, this stuff is also deleted. sounds like it is.

yes @diracdeltas it is deleted or not persisted if the user deletes or disables storage. This is just like any LLM text response, just a little more structured.

petemill commented:
Ran in to an issue that android doesn't have sanitized_image_source.{h,cc} included in the `chrome/browser/ui` target due to a condition. Exploring patching it.

bridiver commented on components/ai_chat/core/browser/ai_chat_database.cc:
```diff
@@ -330,6 +375,34 @@ std::vector<mojom::ConversationTurnPtr> AIChatDatabase::GetConversationEntries(
     }
 
+    // Web Source events
+    {
```
this is a rather large method, can we break it up in a follow-up?

bridiver replied:
and then we can also split up unit testing here as discussed in DM.

bridiver reviewed (approved):
code owner changes ok

mkarolin reviewed (approved):
`strings`++

brave-builds commented:
Released in [v1.77.28](https://github.com/brave/brave-browser/releases/tag/v1.77.28)
